script/src/verify.rs

```diff
@@ -455,6 +467,267 @@ impl<'a, DL: CellDataProvider + HeaderProvider> TransactionScriptsVerifier<'a, D
+    ///
+    /// ## Returns
+    ///
+    /// It returns the total consumed cycles if verification completed,
+    /// If verify is suspended, a snapshot will retruned.
+    pub fn resumable_verify(&self, limit_cycles: Cycle) -> Result<VerifyResult, Error> {
+        let mut cycles = 0;
+
+        let groups: Vec<_> = self.groups().collect();
+        for (idx, (ty, hash, group)) in groups.iter().enumerate() {
+            // vm should early return invalid cycles
+            let step_cycle = limit_cycles.checked_sub(cycles).ok_or_else(|| {
+                ScriptError::VMInternalError(format!(
+                    "expect invalid cycles {} {}",
+                    limit_cycles, cycles
+                ))
+                .source(group)
+            })?;
+
+            match self.verify_group_with_chunk(&group, step_cycle, None) {
+                Ok(ChunkState::Completed(cycle)) => {
+                    cycles = wrapping_cycles_add(cycles, cycle, limit_cycles, group)?;
+                }
+                Ok(ChunkState::VM(vm)) => {
+                    cycles = wrapping_cycles_add(cycles, vm.machine.cycles(), limit_cycles, group)?;
+
+                    let current = (*ty, (*hash).to_owned());
+
+                    let remain = groups
+                        .iter()
+                        .skip(idx)
+                        .map(|(ty, hash, _g)| (*ty, (*hash).to_owned()))
+                        .collect();
+
+                    let snap =
+                        self.make_snapshot(vm, group, current, remain, cycles, limit_cycles)?;
+
+                    return Ok(VerifyResult::Suspended(Arc::new(snap)));
+                }
+                Err(e) => {
+                    return Err(e.source(group).into());
+                }
+            }
+        }
+
+        Ok(VerifyResult::Completed(cycles))
+    }
+
+    /// Resuming an suspended verify
+    ///
+    /// ## Params
+    ///
+    /// * `snap` - Captured transaction verification snapshot.
+    ///
+    /// * `limit_cycles` - Maximium allowed cycles to run the scripts. The verification quits early
+    /// when the consumed cycles exceed the limit.
+    ///
+    /// ## Returns
+    ///
+    /// It returns the total consumed cycles if verification completed,
+    /// If verify is suspended, a new snapshot will retruned.
+    pub fn resume(
+        &self,
+        snap: &TransactionSnapshot,
+        limit_cycles: Cycle,
+    ) -> Result<VerifyResult, Error> {
+        let mut cycles = snap.current_cycles;
+
+        let current_group = self
+            .find_script_group(snap.current.0, &snap.current.1)
+            .ok_or_else(|| {
+                ScriptError::VMInternalError(format!("snapshot group missing {:?}", snap.current))
+                    .unknown_source()
+            })?;
+
+        if limit_cycles < snap.current_cycles {
+            return Err(ScriptError::ExceededMaximumCycles(limit_cycles)
+                .source(current_group)
```

Comment:
Here I think we need to use unknow group, it is not necessarily the problem of the current group